File: flowquery-py/src/parsing/functions/predicate_any.py
"""PredicateAny function."""

from typing import Any

from .function_metadata import FunctionDef
from .predicate_function import PredicateFunction


@FunctionDef({
    "description": "Returns true if at least one element in the list satisfies the condition.",
    "category": "predicate",
    "parameters": [
        {"name": "variable", "description": "Variable name to bind each element", "type": "string"},
        {"name": "list", "description": "List to iterate over", "type": "array"},
        {"name": "where", "description": "Condition to test for each element", "type": "boolean"},
    ],
    "output": {"description": "True if any element satisfies the condition", "type": "boolean", "example": True},
    "examples": [
        "RETURN any(n IN [1, 2, 3] WHERE n > 2)",
    ],
})
class PredicateAny(PredicateFunction):
    def __init__(self) -> None:
        super().__init__("any")

    def value(self) -> Any:
        ref = self.reference
        if hasattr(ref, 'referred'):
            ref.referred = self._value_holder
        array = self.array.value()
        if array is None or not isinstance(array, list):
            return False
        for item in array:
            self._value_holder.holder = item
            if self.where is not None and self.where.value():
                return True
        return False

File: src/parsing/functions/predicate_any.ts
import { FunctionDef } from "./function_metadata";
import PredicateFunction from "./predicate_function";

@FunctionDef({
    description:
        "Returns true if at least one element in the list satisfies the condition. Uses syntax: any(variable IN list WHERE condition)",
    category: "predicate",
    parameters: [
        { name: "variable", description: "Variable name to bind each element", type: "string" },
        { name: "list", description: "List to iterate over", type: "array" },
        {
            name: "where",
            description: "Condition to test for each element",
            type: "boolean",
        },
    ],
    output: {
        description: "True if any element satisfies the condition",
        type: "boolean",
        example: true,
    },
    examples: [
        "RETURN any(n IN [1, 2, 3] WHERE n > 2)",
        "WITH [1, -1, 2] AS nums RETURN any(n IN nums WHERE n < 0)",
    ],
})
class PredicateAny extends PredicateFunction {
    constructor() {
        super("any");
    }

    public value(): any {
        this.reference.referred = this._valueHolder;
        const array = this.array.value();
        if (array === null || !Array.isArray(array)) {
            return false;
        }
        for (let i = 0; i < array.length; i++) {
            this._valueHolder.holder = array[i];
            if (this.where !== null && this.where.value()) {
                return true;
            }
        }
        return false;
    }
}

export default PredicateAny;
